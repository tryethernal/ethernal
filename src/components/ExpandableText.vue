<template>
  <div class="expandable-text">
      <div class="d-block overflow-auto pr-10" style="white-space: pre-wrap;">
        <template v-if="pre">
          <pre style="white-space: pre-wrap;">{{ displayText }}</pre>
        </template>
        <template v-else>
          {{ displayText }}
        </template>
      </div>
      <div class="copy-btn-container">
        <v-btn
          size="small" 
          color="primary"
          variant="text" 
          @click="copyToClipboard" 
          v-tooltip="'Copy to clipboard'"
        >
          <v-icon>{{ hasCopied ? 'mdi-check' : 'mdi-content-copy' }}</v-icon>
        </v-btn>
      </div>
    
    <!-- Toggle button -->
    <div v-if="needsTruncation" class="text-center mt-2">
      <a href="#" class="no-decoration text-uppercase d-flex align-center justify-center" @click.prevent="toggleExpand">
        <v-icon size="small" class="mr-1">mdi-eye{{ isExpanded ? '-off' : '' }}</v-icon>
        {{ isExpanded ? 'VIEW LESS' : 'VIEW ALL' }}
      </a>
    </div>
    <!-- Hidden input for fallback copy method -->
    <input type="hidden" ref="copyElement" :value="rawText">
  </div>
</template>

<script setup>
import { ref, computed, useSlots } from 'vue';

const props = defineProps({
  text: {
    type: String,
    default: ''
  },
  maxChars: {
    type: Number,
    default: 500
  },
  pre: {
    type: Boolean,
    default: false
  }
});

// Emit events when expansion state changes
const emit = defineEmits(['expand', 'collapse']);

const slots = useSlots();
const isExpanded = ref(false);
const hasCopied = ref(false);
const copyElement = ref(null);

const displayText = computed(() => {
  if (isExpanded.value || rawText.value.length <= props.maxChars) {
    return rawText.value;
  }
  return rawText.value.substring(0, props.maxChars) + '...';
});

const rawText = computed(() => {
  return slots.default ? slots.default()[0].children : props.text;
});

const needsTruncation = computed(() => {
  return rawText.value.length > props.maxChars;
});

const toggleExpand = () => {
  isExpanded.value = !isExpanded.value;
  
  // Emit appropriate event
  if (isExpanded.value) {
    emit('expand');
  } else {
    emit('collapse');
  }
};

const copyToClipboard = async () => {
    // Fall back to execCommand method if navigator.clipboard fails
    try {
      const element = copyElement.value;
      element.setAttribute('type', 'text');
      element.select();
      document.execCommand('copy');
      hasCopied.value = true;
      setTimeout(() => {
        hasCopied.value = false;
      }, 2000);
    } catch (fallbackError) {
      console.error('Failed to copy text with fallback method: ', fallbackError);
    } finally {
      if (copyElement.value) {
        copyElement.value.setAttribute('type', 'hidden');
        window.getSelection().removeAllRanges();
      }
    }
};
</script>

<style scoped>
.expandable-text {
  position: relative;
  margin-bottom: 4px;
  background-color: var(--card-background);
}

.copy-btn-container {
  position: absolute;
  top: 0px;
  right: 0px;
}

.no-decoration {
  text-decoration: none;
}

pre {
  background-color: var(--card-background);
  margin: 0;
  padding: 8px;
  font-family: monospace;
}

/* Dark mode support */
:deep(.v-theme--dark) {
  .expandable-text,
  pre {
    background-color: var(--card-background);
  }
}
</style>
