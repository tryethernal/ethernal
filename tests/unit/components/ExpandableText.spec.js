import ExpandableText from '@/components/ExpandableText.vue';

describe('ExpandableText.vue', () => {
  const shortText = 'Short text';
  const longText = 'a'.repeat(600); // Text longer than default maxChars (500)

  it('should render short text without truncation', async () => {
    const wrapper = mount(ExpandableText, {
      props: {
        text: shortText
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render truncated long text with view more button', async () => {
    const wrapper = mount(ExpandableText, {
      props: {
        text: longText
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should expand text when clicking view more', async () => {
    const wrapper = mount(ExpandableText, {
      props: {
        text: longText
      }
    });

    await wrapper.find('a').trigger('click');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render text in pre tag when pre prop is true', async () => {
    const wrapper = mount(ExpandableText, {
      props: {
        text: shortText,
        pre: true
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should show copy success icon after copying', async () => {
    // Setup fake timers
    vi.useFakeTimers();
    
    // Mock execCommand since it's not available in test environment
    document.execCommand = vi.fn();
    
    const wrapper = mount(ExpandableText, {
      props: {
        text: shortText
      }
    });

    await wrapper.find('.copy-btn-container .v-btn').trigger('click');
    expect(wrapper.html()).toMatchSnapshot();

    // Advance timer to simulate waiting for icon to revert
    await vi.advanceTimersByTime(2100);
    expect(wrapper.html()).toMatchSnapshot();

    // Cleanup
    vi.useRealTimers();
  });

  it('should emit expand/collapse events when toggling', async () => {
    const wrapper = mount(ExpandableText, {
      props: {
        text: longText
      }
    });

    await wrapper.find('a').trigger('click');
    expect(wrapper.emitted()).toHaveProperty('expand');

    await wrapper.find('a').trigger('click');
    expect(wrapper.emitted()).toHaveProperty('collapse');
  });

  it('should handle slot content', async () => {
    const wrapper = mount(ExpandableText, {
      slots: {
        default: 'Slot content'
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
}); 