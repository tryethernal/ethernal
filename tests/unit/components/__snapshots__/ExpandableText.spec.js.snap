// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExpandableText.vue > should expand text when clicking view more 1`] = `
"<div data-v-130fe57e="" class="expandable-text">
  <div data-v-130fe57e="" class="d-block overflow-auto pr-10" style="white-space: pre-wrap;">aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa</div>
  <div data-v-130fe57e="" class="copy-btn-container"><button data-v-130fe57e="" type="button" class="v-btn v-theme--light text-primary v-btn--density-default v-btn--size-small v-btn--variant-text" aria-describedby="v-tooltip-1"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
      <!----><span class="v-btn__content" data-no-activator=""><i data-v-130fe57e="" class="mdi-content-copy mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span>
      <!---->
      <!---->
      <!---->
      <!--teleport start-->
      <!--teleport end-->
    </button></div><!-- Toggle button -->
  <div data-v-130fe57e="" class="text-center mt-2"><a data-v-130fe57e="" href="#" class="no-decoration text-uppercase d-flex align-center justify-center"><i data-v-130fe57e="" class="mdi-eye-off mdi v-icon notranslate v-theme--light v-icon--size-small mr-1" aria-hidden="true"></i> VIEW LESS</a></div><!-- Hidden input for fallback copy method --><input data-v-130fe57e="" type="hidden" value="aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa">
</div>"
`;

exports[`ExpandableText.vue > should handle slot content 1`] = `
"<div data-v-130fe57e="" class="expandable-text">
  <div data-v-130fe57e="" class="d-block overflow-auto pr-10" style="white-space: pre-wrap;">Slot content</div>
  <div data-v-130fe57e="" class="copy-btn-container"><button data-v-130fe57e="" type="button" class="v-btn v-theme--light text-primary v-btn--density-default v-btn--size-small v-btn--variant-text"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
      <!----><span class="v-btn__content" data-no-activator=""><i data-v-130fe57e="" class="mdi-content-copy mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span>
      <!---->
      <!---->
      <!---->
      <!--teleport start-->
      <!--teleport end-->
    </button></div><!-- Toggle button -->
  <!--v-if-->
  <!-- Hidden input for fallback copy method --><input data-v-130fe57e="" type="hidden" value="Slot content">
</div>"
`;

exports[`ExpandableText.vue > should render short text without truncation 1`] = `
"<div data-v-130fe57e="" class="expandable-text">
  <div data-v-130fe57e="" class="d-block overflow-auto pr-10" style="white-space: pre-wrap;">Short text</div>
  <div data-v-130fe57e="" class="copy-btn-container"><button data-v-130fe57e="" type="button" class="v-btn v-theme--light text-primary v-btn--density-default v-btn--size-small v-btn--variant-text"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
      <!----><span class="v-btn__content" data-no-activator=""><i data-v-130fe57e="" class="mdi-content-copy mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span>
      <!---->
      <!---->
      <!---->
      <!--teleport start-->
      <!--teleport end-->
    </button></div><!-- Toggle button -->
  <!--v-if-->
  <!-- Hidden input for fallback copy method --><input data-v-130fe57e="" type="hidden" value="Short text">
</div>"
`;

exports[`ExpandableText.vue > should render text in pre tag when pre prop is true 1`] = `
"<div data-v-130fe57e="" class="expandable-text">
  <div data-v-130fe57e="" class="d-block overflow-auto pr-10" style="white-space: pre-wrap;"><pre data-v-130fe57e="" style="white-space: pre-wrap;">Short text</pre>
  </div>
  <div data-v-130fe57e="" class="copy-btn-container"><button data-v-130fe57e="" type="button" class="v-btn v-theme--light text-primary v-btn--density-default v-btn--size-small v-btn--variant-text"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
      <!----><span class="v-btn__content" data-no-activator=""><i data-v-130fe57e="" class="mdi-content-copy mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span>
      <!---->
      <!---->
      <!---->
      <!--teleport start-->
      <!--teleport end-->
    </button></div><!-- Toggle button -->
  <!--v-if-->
  <!-- Hidden input for fallback copy method --><input data-v-130fe57e="" type="hidden" value="Short text">
</div>"
`;

exports[`ExpandableText.vue > should render truncated long text with view more button 1`] = `
"<div data-v-130fe57e="" class="expandable-text">
  <div data-v-130fe57e="" class="d-block overflow-auto pr-10" style="white-space: pre-wrap;">aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa...</div>
  <div data-v-130fe57e="" class="copy-btn-container"><button data-v-130fe57e="" type="button" class="v-btn v-theme--light text-primary v-btn--density-default v-btn--size-small v-btn--variant-text"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
      <!----><span class="v-btn__content" data-no-activator=""><i data-v-130fe57e="" class="mdi-content-copy mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span>
      <!---->
      <!---->
      <!---->
      <!--teleport start-->
      <!--teleport end-->
    </button></div><!-- Toggle button -->
  <div data-v-130fe57e="" class="text-center mt-2"><a data-v-130fe57e="" href="#" class="no-decoration text-uppercase d-flex align-center justify-center"><i data-v-130fe57e="" class="mdi-eye mdi v-icon notranslate v-theme--light v-icon--size-small mr-1" aria-hidden="true"></i> VIEW ALL</a></div><!-- Hidden input for fallback copy method --><input data-v-130fe57e="" type="hidden" value="aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa">
</div>"
`;

exports[`ExpandableText.vue > should show copy success icon after copying 1`] = `
"<div data-v-130fe57e="" class="expandable-text">
  <div data-v-130fe57e="" class="d-block overflow-auto pr-10" style="white-space: pre-wrap;">Short text</div>
  <div data-v-130fe57e="" class="copy-btn-container"><button data-v-130fe57e="" type="button" class="v-btn v-theme--light text-primary v-btn--density-default v-btn--size-small v-btn--variant-text" aria-describedby="v-tooltip-1"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
      <!----><span class="v-btn__content" data-no-activator=""><i data-v-130fe57e="" class="mdi-check mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span>
      <!---->
      <!---->
      <!---->
      <!--teleport start-->
      <!--teleport end-->
    </button></div><!-- Toggle button -->
  <!--v-if-->
  <!-- Hidden input for fallback copy method --><input data-v-130fe57e="" type="hidden" value="Short text">
</div>"
`;

exports[`ExpandableText.vue > should show copy success icon after copying 2`] = `
"<div data-v-130fe57e="" class="expandable-text">
  <div data-v-130fe57e="" class="d-block overflow-auto pr-10" style="white-space: pre-wrap;">Short text</div>
  <div data-v-130fe57e="" class="copy-btn-container"><button data-v-130fe57e="" type="button" class="v-btn v-theme--light text-primary v-btn--density-default v-btn--size-small v-btn--variant-text" aria-describedby="v-tooltip-1"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
      <!----><span class="v-btn__content" data-no-activator=""><i data-v-130fe57e="" class="mdi-content-copy mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span>
      <!---->
      <!---->
      <!---->
      <!--teleport start-->
      <!--teleport end-->
    </button></div><!-- Toggle button -->
  <!--v-if-->
  <!-- Hidden input for fallback copy method --><input data-v-130fe57e="" type="hidden" value="Short text">
</div>"
`;
